parsec/backend/postgresql/pki.py

```diff
@@ -0,0 +1,439 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+
+import hashlib
+from typing import List
+from uuid import UUID
+from pendulum import DateTime
+
+from parsec.api.protocol import OrganizationID
+from parsec.api.protocol.pki import PkiEnrollmentStatus
+from parsec.api.protocol.types import UserID, UserProfile
+from parsec.backend.backend_events import BackendEvent
+from parsec.backend.postgresql.handler import send_signal
+from parsec.backend.user import UserActiveUsersLimitReached, UserAlreadyExistsError
+from parsec.backend.user_type import User, Device
+from parsec.backend.pki import (
+    PkiEnrollmentActiveUsersLimitReached,
+    PkiEnrollmentAlreadyEnrolledError,
+    PkiEnrollmentAlreadyExistError,
+    PkiEnrollmentCertificateAlreadySubmittedError,
+    PkiEnrollmentIdAlreadyUsedError,
+    PkiEnrollmentInfo,
+    PkiEnrollmentInfoAccepted,
+    PkiEnrollmentInfoCancelled,
+    PkiEnrollmentInfoRejected,
+    PkiEnrollmentInfoSubmitted,
+    PkiEnrollmentListItem,
+    BasePkiEnrollmentComponent,
+    PkiEnrollmentNoLongerAvailableError,
+    PkiEnrollmentNotFoundError,
+)
+from parsec.backend.postgresql import PGHandler
+from parsec.backend.postgresql.utils import Q, q_organization_internal_id, q_device_internal_id
+from parsec.backend.postgresql.user_queries.create import q_create_user
+
+_q_get_pki_enrollment_from_certificate_sha1 = Q(
+    f"""
+    SELECT * FROM pki_enrollment
+    WHERE (
+        organization = { q_organization_internal_id("$organization_id") }
+        AND submitter_der_x509_certificate_sha1=$submitter_der_x509_certificate_sha1
+    )
+    ORDER BY enrollment_state ASC
```

Comment:
Can we order by `_id` instead ? I feel ordering by state is not really clear, and I may break if we add more states...

This is especially true given the comment where this query is used say we are retrieving the last attempt ;-)

Reply:
And we may want to rename this query `_q_get_last_pki_enrollment_from_certificate_sha1` and add a `LIMIT 1` to the SQL